eclipse-theia/theia
# Implement Test API

#### What it does

Closes https://github.com/eclipse-theia/theia/issues/10669

Implements support for the VS Code testing API in the usual fashion: there is a test service interface and implementation that is plugin-agnostic (`packages/test?`) and a plugin-specific implementation in the plugin-ext package.

#### How to test
There is a sample test controller contribution active in the sample applications via the "api-samples" folder. The test controller shows up in the Test Explorer View. You'll need to invoke the "Add Some Tests" command to populate the controller with tests.
In addition, I used the https://open-vsx.org/extension/ms-playwright/playwright extension to test the code on the Theia source code itself. On Windows, you'll need to fix the playwright "theia:start" yarn task like so:
```
rimraf .tmp.cfg && set THEIA_CONFIG_DIR=./.tmp.cfg  && yarn --cwd ../browser start
```

Note that this is minimal implementation that lets us satisfy the test API. It consists of "Test Exlorer" and "Test Result" views showing up in the left sidebar and a "Test Output" and a "Test Result" view showing up at the bottom. You'll have to open those second two views manually. 

#### Follow-ups
While the API is complete, there are a couple of things which the UI will never use:

1. No support for "continous run". 
Even if a test plugin supports continoous runs, the UI will never invoke a test run in this fashion.
2. No support for `invalidateTestResults`. 
The UI has no support for showing test results as outdated, so that API has no function
3. No Test Markers
We show no marker annotations for tests in the editor. To run tests, you'll need to use the UI in the Test Explorer view.
4. Build out convenience functions in the UI, like
   * automatically open "Test Output" and "Test Result" views on double click
   * be smarter on what to show in the output and results: aggregate data if non available on the selection
5. No support for "coverage" test profiles
Since the whole coverage API is in "proposed" state, it is not implemented.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
